Fix #46. Add ability to create Web and Function Apps from Azure Explorer

sdubov opened the pull request:
Fix #87
Fix #46

* Refactor actions logic to create new Azure entries
* Add a new action to create a Web App
* Add a new action to create a Function App
* Add new icons for Azure Explorer action for adding a new Azure Entries
* Fix issues in Storage Account radio button groups (existing/create new)
* Refactor Create SQL Server and Create SQL Database actions and separate UI components from main dialog logic
* Minor updates in progress indicator text

<img width="278" alt="CreateFunctionApp_Context" src="https://user-images.githubusercontent.com/6064345/91719130-6aa7dc00-eb9d-11ea-85db-d11181a4dc66.png">

<img width="567" alt="CreateFunctionApp_Dialog" src="https://user-images.githubusercontent.com/6064345/91719136-6e3b6300-eb9d-11ea-9b23-a32df6fbc36a.png">

<img width="278" alt="CreateFunctionApp_Context" src="https://user-images.githubusercontent.com/6064345/91719147-74314400-eb9d-11ea-8c46-e0b907752368.png">

<img width="479" alt="CreateWebApp_Dialog" src="https://user-images.githubusercontent.com/6064345/91719156-77c4cb00-eb9d-11ea-8466-8ff95ae3be4c.png">

sdubov committed:
Fix #46. Add ability to create Web and Function Apps from Azure Explorer

- Refactor actions logic to create new Azure entries
- Add a new action to create a Web App
- Add a new action to create a Function App
- Add new icons for Azure Explorer action for adding a new Azure Entries
- Fix issues in Storage Account radio button groups (existing/create new)
- Refactor Create SQL Server and Create SQL Database actions and separate UI components from main dialog logic
- Minor updates in progress indicator text

maartenba commented on PluginsAndFeatures/azure-toolkit-for-intellij/rider/resources/messages/RiderAzureMessages.properties:
```diff
@@ -336,6 +336,8 @@ process_event.publish.web_apps.get_existing=Got existing Web App with Id: ''{0}'
 process_event.publish.web_apps.id_not_defined=Web App ID is not defined
 process_event.publish.web_apps.name_not_defined=Web App name not provided
 process_event.publish.web_apps.set_startup_file=Set Startup File for a web app ''{0}'' to: ''{1}''
+process_event.publish.web_apps.windows.net_framework_not_defined=Net Framework not provided
```
```suggestion
process_event.publish.web_apps.windows.net_framework_not_defined=.NET Framework not provided
```

maartenba commented on PluginsAndFeatures/azure-toolkit-for-intellij/rider/resources/messages/RiderAzureMessages.properties:
```diff
@@ -336,6 +336,8 @@ process_event.publish.web_apps.get_existing=Got existing Web App with Id: ''{0}'
 process_event.publish.web_apps.name_not_defined=Web App name not provided
 process_event.publish.web_apps.set_startup_file=Set Startup File for a web app ''{0}'' to: ''{1}''
+process_event.publish.web_apps.windows.net_framework_not_defined=Net Framework not provided
+process_event.publish.web_apps.windows.net_core_runtime_not_defined=Net Core Runtime not provided
```
```suggestion
process_event.publish.web_apps.windows.net_core_runtime_not_defined=.NET Core Runtime not provided
```

maartenba reviewed (changes requested):
Looks good @sdubov, although I ran into a few issues:

* Create Web App - App Service plan region does not seem to be respected. I picked West Europe and got a deployment in Australia.
* Create function app - Getting the following exception:

```
com.microsoft.azure.CloudException: Status code 404, {"error":{"code":"SubscriptionNotFound","message":"The subscription 'providers' could not be found."}}: The subscription 'providers' could not be found.
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
	at com.microsoft.rest.ServiceResponseBuilder.build(ServiceResponseBuilder.java:122)
	at com.microsoft.azure.AzureResponseBuilder.build(AzureResponseBuilder.java:56)
	at com.microsoft.azure.management.storage.implementation.StorageAccountsInner.checkNameAvailabilityDelegate(StorageAccountsInner.java:223)
	at com.microsoft.azure.management.storage.implementation.StorageAccountsInner.access$000(StorageAccountsInner.java:57)
	at com.microsoft.azure.management.storage.implementation.StorageAccountsInner$2.call(StorageAccountsInner.java:210)
	at com.microsoft.azure.management.storage.implementation.StorageAccountsInner$2.call(StorageAccountsInner.java:206)
	at rx.internal.operators.OnSubscribeMap$MapSubscriber.onNext(OnSubscribeMap.java:69)
	at retrofit2.adapter.rxjava.CallArbiter.deliverResponse(CallArbiter.java:120)
	at retrofit2.adapter.rxjava.CallArbiter.emitResponse(CallArbiter.java:102)
	at retrofit2.adapter.rxjava.CallExecuteOnSubscribe.call(CallExecuteOnSubscribe.java:46)
	at retrofit2.adapter.rxjava.CallExecuteOnSubscribe.call(CallExecuteOnSubscribe.java:24)
	at rx.Observable.unsafeSubscribe(Observable.java:10327)
	at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:48)
	at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:33)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:48)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30)
	at rx.Observable.unsafeSubscribe(Observable.java:10327)
	at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:48)
	at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:33)
	at rx.Observable.unsafeSubscribe(Observable.java:10327)
	at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:48)
	at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:33)
	at rx.Observable.unsafeSubscribe(Observable.java:10327)
	at rx.internal.operators.DeferredScalarSubscriber.subscribeTo(DeferredScalarSubscriber.java:153)
	at rx.internal.operators.OnSubscribeTakeLastOne.call(OnSubscribeTakeLastOne.java:32)
	at rx.internal.operators.OnSubscribeTakeLastOne.call(OnSubscribeTakeLastOne.java:22)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:48)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30)
	at rx.Observable.subscribe(Observable.java:10423)
	at rx.Observable.subscribe(Observable.java:10390)
	at rx.observables.BlockingObservable.blockForSingle(BlockingObservable.java:443)
	at rx.observables.BlockingObservable.last(BlockingObservable.java:226)
	at com.microsoft.azure.management.storage.implementation.StorageAccountsImpl.checkNameAvailability(StorageAccountsImpl.java:41)
	at com.microsoft.azuretools.core.mvp.model.storage.AzureStorageAccountMvpModel.checkNameAvailability(AzureStorageAccountMvpModel.kt:85)
	at com.microsoft.azuretools.core.mvp.model.storage.AzureStorageAccountMvpModel.isStorageAccountNameExist(AzureStorageAccountMvpModel.kt:95)
	at com.microsoft.intellij.helpers.validator.StorageAccountValidator.isStorageAccountExist(StorageAccountValidator.kt:94)
	at com.microsoft.intellij.helpers.validator.StorageAccountValidator.checkStorageAccountNameExists(StorageAccountValidator.kt:87)
	at com.microsoft.intellij.helpers.validator.StorageAccountValidator.validateStorageAccountName(StorageAccountValidator.kt:46)
	at com.microsoft.intellij.ui.component.StorageAccountSelector.validateComponent(StorageAccountSelector.kt:101)
	at com.microsoft.intellij.ui.forms.appservice.functionapp.CreateFunctionAppDialog.validateComponent(CreateFunctionAppDialog.kt:162)
	at com.microsoft.intellij.ui.forms.base.AzureCreateDialogBase.doValidateAll(AzureCreateDialogBase.kt:53)
	at com.intellij.openapi.ui.DialogWrapper$OkAction.doAction(DialogWrapper.java:1869)
	at com.intellij.openapi.ui.DialogWrapper$DialogWrapperAction.actionPerformed(DialogWrapper.java:1836)
	at java.desktop/javax.swing.AbstractButton.fireActionPerformed(AbstractButton.java:1967)
	at java.desktop/javax.swing.AbstractButton$Handler.actionPerformed(AbstractButton.java:2308)
	at java.desktop/javax.swing.DefaultButtonModel.fireActionPerformed(DefaultButtonModel.java:405)
	at java.desktop/javax.swing.DefaultButtonModel.setPressed(DefaultButtonModel.java:262)
	at java.desktop/javax.swing.plaf.basic.BasicButtonListener.mouseReleased(BasicButtonListener.java:270)
	at java.desktop/java.awt.Component.processMouseEvent(Component.java:6650)
	at java.desktop/javax.swing.JComponent.processMouseEvent(JComponent.java:3345)
	at java.desktop/java.awt.Component.processEvent(Component.java:6415)
	at java.desktop/java.awt.Container.processEvent(Container.java:2263)
	at java.desktop/java.awt.Component.dispatchEventImpl(Component.java:5025)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2321)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4857)
	at java.desktop/java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4918)
	at java.desktop/java.awt.LightweightDispatcher.processMouseEvent(Container.java:4547)
	at java.desktop/java.awt.LightweightDispatcher.dispatchEvent(Container.java:4488)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2307)
	at java.desktop/java.awt.Window.dispatchEventImpl(Window.java:2773)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4857)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:778)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:727)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:95)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:751)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:749)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:748)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.java:967)
	at com.intellij.ide.IdeEventQueue.dispatchMouseEvent(IdeEventQueue.java:904)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:836)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$8(IdeEventQueue.java:450)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computePrioritized(CoreProgressManager.java:744)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$9(IdeEventQueue.java:449)
	at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:802)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:503)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:117)
	at java.desktop/java.awt.WaitDispatchSupport$2.run(WaitDispatchSupport.java:190)
	at java.desktop/java.awt.WaitDispatchSupport$4.run(WaitDispatchSupport.java:235)
	at java.desktop/java.awt.WaitDispatchSupport$4.run(WaitDispatchSupport.java:233)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.desktop/java.awt.WaitDispatchSupport.enter(WaitDispatchSupport.java:233)
	at java.desktop/java.awt.Dialog.show(Dialog.java:1063)
	at com.intellij.openapi.ui.impl.DialogWrapperPeerImpl$MyDialog.show(DialogWrapperPeerImpl.java:710)
	at com.intellij.openapi.ui.impl.DialogWrapperPeerImpl.show(DialogWrapperPeerImpl.java:437)
	at com.intellij.openapi.ui.DialogWrapper.doShow(DialogWrapper.java:1702)
	at com.intellij.openapi.ui.DialogWrapper.show(DialogWrapper.java:1661)
	at com.microsoft.intellij.serviceexplorer.azure.functionapp.actions.FunctionAppCreateAction.actionPerformed(FunctionAppCreateAction.kt:60)
	at com.microsoft.tooling.msservices.serviceexplorer.NodeActionListener.actionPerformedAsync(NodeActionListener.java:86)
	at com.microsoft.tooling.msservices.serviceexplorer.NodeAction.fireNodeActionEvent(NodeAction.java:61)
	at com.microsoft.intellij.components.ServerExplorerToolWindowFactory.lambda$createPopupMenuForNode$1(ServerExplorerToolWindowFactory.java:216)
	at java.desktop/javax.swing.AbstractButton.fireActionPerformed(AbstractButton.java:1967)
	at java.desktop/javax.swing.AbstractButton$Handler.actionPerformed(AbstractButton.java:2308)
	at java.desktop/javax.swing.DefaultButtonModel.fireActionPerformed(DefaultButtonModel.java:405)
	at java.desktop/javax.swing.DefaultButtonModel.setPressed(DefaultButtonModel.java:262)
	at java.desktop/javax.swing.AbstractButton.doClick(AbstractButton.java:369)
	at java.desktop/javax.swing.plaf.basic.BasicMenuItemUI.doClick(BasicMenuItemUI.java:980)
	at java.desktop/javax.swing.plaf.basic.BasicMenuItemUI$Handler.mouseReleased(BasicMenuItemUI.java:1024)
	at java.desktop/java.awt.Component.processMouseEvent(Component.java:6650)
	at java.desktop/javax.swing.JComponent.processMouseEvent(JComponent.java:3345)
	at java.desktop/java.awt.Component.processEvent(Component.java:6415)
	at java.desktop/java.awt.Container.processEvent(Container.java:2263)
	at java.desktop/java.awt.Component.dispatchEventImpl(Component.java:5025)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2321)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4857)
	at java.desktop/java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4918)
	at java.desktop/java.awt.LightweightDispatcher.processMouseEvent(Container.java:4547)
	at java.desktop/java.awt.LightweightDispatcher.dispatchEvent(Container.java:4488)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2307)
	at java.desktop/java.awt.Window.dispatchEventImpl(Window.java:2773)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4857)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:778)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:727)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:95)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:751)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:749)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:748)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.java:967)
	at com.intellij.ide.IdeEventQueue.dispatchMouseEvent(IdeEventQueue.java:904)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:836)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$8(IdeEventQueue.java:450)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computePrioritized(CoreProgressManager.java:744)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$9(IdeEventQueue.java:449)
	at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:802)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:503)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
```

sdubov committed:
Fix PR comments

- Fix exception on creating Function app
- Fix text for .NET runtime
- Remove existence check from UI component validation logic and left it for a caller side

sdubov commented:
@maartenba, thank you for checking. 

1. Could you please provide more details in repro steps for "Create Web App (location)" issue? I checked for both new and existing App Service Plan and Web Apps are created with correct location on my side.
2. Fixed the exception in creating storage account when adding a function app.
3. Fixed text suggestions.

maartenba commented:
Thanks! Web app now seems to deploy in the correct location as well, so that's fixed :)

Functions: still getting an exception:

![image](https://user-images.githubusercontent.com/485230/91800058-9ae88c80-ec28-11ea-9038-0ffea81ee0c5.png)

On creation:

```
java.security.InvalidParameterException: The specified ID `` is not a valid Azure resource ID.
	at com.microsoft.azure.management.resources.fluentcore.arm.ResourceId.<init>(ResourceId.java:40)
	at com.microsoft.azure.management.resources.fluentcore.arm.ResourceId.fromString(ResourceId.java:112)
	at com.microsoft.azure.management.resources.fluentcore.arm.collection.implementation.GroupableResourcesImpl.getByIdAsync(GroupableResourcesImpl.java:74)
	at com.microsoft.azure.management.resources.fluentcore.arm.collection.implementation.GroupableResourcesImpl.getById(GroupableResourcesImpl.java:69)
	at com.microsoft.azure.management.resources.fluentcore.arm.collection.implementation.GroupableResourcesImpl.getById(GroupableResourcesImpl.java:34)
	at com.microsoft.azuretools.core.mvp.model.storage.AzureStorageAccountMvpModel.getStorageAccountById(AzureStorageAccountMvpModel.kt:81)
	at com.microsoft.azuretools.core.mvp.model.functionapp.AzureFunctionAppMvpModel.createFunctionApp(AzureFunctionAppMvpModel.kt:171)
	at com.microsoft.intellij.ui.forms.appservice.functionapp.CreateFunctionAppDialog$doOKAction$1.run(CreateFunctionAppDialog.kt:117)
	at com.intellij.openapi.progress.impl.CoreProgressManager$TaskRunnable.run(CoreProgressManager.java:935)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcessWithProgressAsync$5(CoreProgressManager.java:442)
	at com.intellij.openapi.progress.impl.ProgressRunner.lambda$submit$3(ProgressRunner.java:235)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$2(CoreProgressManager.java:170)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:629)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:581)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:60)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:157)
	at com.intellij.openapi.progress.impl.ProgressRunner.lambda$submit$4(ProgressRunner.java:235)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1700)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:668)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:665)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1.run(Executors.java:665)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

sdubov committed:
Fix Storage Account create logic and validation for Azure components

sdubov commented:
Wow... Nice catch! Fixed.

maartenba reviewed (approved):
Works now, woohoo!

sdubov commented:
@maartenba, thank you!
